packages/node/src/node.ts

```diff
@@ -101,58 +111,56 @@ export class Node {
             "Reached OP_SIGN middleware without generated commitment."
           );
         }
-        context.signature = {
-          v: -1,
-          r: HashZero,
-          s: HashZero
-        };
-        // TODO: Replace the above hack with the below
-        // context.signature = this.signer.signDigest(
-        //   context.commitment.hashToSign()
-        // );
+
+        context.signature = this.signer.signDigest(
+          context.commitment.hashToSign()
+        );
+
         next();
       }
     );
-  }
 
-  private registerIoMiddleware() {
-    this.instructionExecutor.register(
+    instructionExecutor.register(
       Opcode.IO_SEND,
       async (message: ProtocolMessage, next: Function, context: Context) => {
-        await this.messagingService.send(context.outbox[0].toAddress, {
-          from: this.address,
-          event: NODE_EVENTS.PROTOCOL_MESSAGE_EVENT,
-          data: context.outbox[0]
-        });
+        const from = getAddress(this.address);
+        const to = getAddress(context.outbox[0].toAddress);
+        const data = context.outbox[0];
+
+        if (context.outbox[0].seq === 1) {
+          this.ioSendDeferrals[to] = new Deferred<
+            NodeMessageWrappedProtocolMessage
+          >();
+        }
+
+        await this.messagingService.send(to, {
+          from,
+          data,
+          event: NODE_EVENTS.PROTOCOL_MESSAGE_EVENT
+        } as NodeMessageWrappedProtocolMessage);
+
         next();
       }
     );
 
-    // TODO: Currently this mocks the response for all protocols as the below
-    //       mocked ProtocolMessage object. Next task is to listen for a real
-    //       message from a counterparty.
-    this.instructionExecutor.register(
+    instructionExecutor.register(
       Opcode.IO_WAIT,
-      (message: ProtocolMessage, next: Function, context: Context) => {
-        // FIXME: This is a temporary hack to get IO_WAIT to progress
-        context.inbox.push({
-          protocol: Protocol.Setup,
-          fromAddress: AddressZero,
-          toAddress: AddressZero,
-          seq: 2,
-          params: {
-            initiatingAddress: AddressZero,
-            respondingAddress: AddressZero,
-            multisigAddress: AddressZero
-          },
-          signature: {
-            v: -1,
-            r: HashZero,
-            s: HashZero
+      async (message: ProtocolMessage, next: Function, context: Context) => {
+        await new Promise(async (resolve, reject) => {
+          try {
+            const checksumAddress = getAddress(message.toAddress);
+            const msg = await this.ioSendDeferrals[checksumAddress].promise;
+            context.inbox.push(msg.data);
+            resolve();
+            next();
+          } catch (e) {
+            reject(e);
```

Comment:
Perhaps this error can be a bit more descriptive to provide context about the failure